src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -421,83 +481,64 @@ protected internal void OnSolutionReloaded(SolutionInfo reloadedSolutionInfo)
         /// </summary>
         protected internal void OnSolutionRemoved()
         {
+            // This currently doesn't use the SetCurrentSolution(transform) pattern as it changes mutable state (open
+            // docs), and as such needs to atomically change both that and the solution-snapshot.
```

Comment:
here's an example of something that could not move over.  Not sure if we can think of a good pattern for this.  the best i can think of is if we change SetCurrentSolution(Transform) to not use optimistic concurrency, and instead just take a fat lock around everything it does. in that case, it would then be safe to do other mutations knowing it would all be atomic.

Reply:
I'm not actually sure the optimistic concurrency buys us anything, other than potential protection against calling other code under a lock and risking deadlock that way.  Practically, at least in VS for Windows, we're taking an async lock around most of this. When I had mistakes where we accidentally took a sync lock, we saw thread pool starvation, which in the optimistic concurrency case would instead have meant we would have had a bunch of threads all spinning against each other.